web/src/components/pages/contribution/listen/listen.tsx

```diff
@@ -128,10 +131,38 @@ class ListenPage extends React.Component<Props, State> {
 
   private vote = (isValid: boolean) => {
     const { clips } = this.state;
+    const { achievement, addNotification } = this.props;
     const clipIndex = this.getClipIndex();
 
     this.stop();
     this.props.vote(isValid, this.state.clips[this.getClipIndex()].id);
+    if (achievement) {
+      addNotification(
+        <div className="achievement">
+          <img
+            src={require('../../dashboard/challenge/images/star.svg')}
+            alt=""
+          />
+          <p className="score">+ 50 points</p>
+          <p>You're on your way! Congrats on your first contribution. </p>
+        </div>
+      );
+      if (Boolean(sessionStorage.getItem('first'))) {
```

Comment:
Here is our thoughts on this after some discussion today:
when clicking "_Copy link_", frontend sends a request to `xxx/challenge/{challenge_token}/achievement/invite`, backend returns `firstInvite: true `if such an invitation is the **first invitation** , AND THEN inserts such an achievement earning in earn table, otherwise, backend returns  `firstInvite: false`.
In the same request, backend also returns `hasAchieved: true` if user already earned the **same-session achievement**;
otherwise, backend returns `hasAchieved: false` if user have not earned the **same-session achievement**, but do NOT insert such an achievement as an earning.


After showing up the **same-session achievement** toast, frontend sends a request to `xxx/challenge/{challenge_token}/achievement/session` just to insert the **same-session achievement** earning in earn table.

In contribution's 2 backend APIs,` saveClip()` and `saveClipVote(),`, backend also piggy-back the `hasAchieved: bool ` to help decide whether or not to save the flag in session storage for the **same-session achievement**.

Reply:
This sounds great! And thank you for the write-up, it helps to read it here before diving into the code. Really appreciate the communication on this.

> After showing up the same-session achievement toast, frontend sends a request to xxx/challenge/{challenge_token}/achievement/session just to insert the same-session achievement earning in earn table.

I think it would be better to save the achievement immediately, like the `firstInvite` achievement. That way, even if there's a bug on the front-end or the toast doesn't fire for some reason, we're still saving the achievement. I think it'll also make cheating-prevention simpler.

I'm going to look into the code now. Excited to see the changes, thanks again!

Reply:
The reason for a separate request to save the achievement is because the backend has no way to know when to save the achievement otherwise. Backend don't have the notion of session. I remember there are some place that keeps changing the session id.
Of course, backend checks to make sure that no such achievement saved twice for one user.